File: ios/Classes/IzipayPaymentPlugin.swift
import Flutter
import UIKit
import IzipayPayButtonSDK
import RLTMXProfiling
import RLTMXProfilingConnections
import VisaSensoryBranding
import MastercardSonic

public class IzipayPaymentPlugin: NSObject, FlutterPlugin, IzipayPaymentDelegate {
  var transactionId : String?
    var transactionStartDatetime : String?
    private var channelResult: FlutterResult!
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "izipay_payment_flutter", binaryMessenger: registrar.messenger())
        let instance = IzipayPaymentPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        channelResult = result
        if (call.method == "startPayment") {
            transactionStartDatetime = getCurrentTimestamp()
            let arguments = call.arguments as! [String: Any]
            let random = Int.random(in: 356547157..<366547157) //solo es un ejemplo, el transactionId debe ser unico y el comercio elige la forma de generarlo
            
            let tranxId = "\(random)"
            self.transactionId = tranxId
            showIzipayForm(
                listPayMethod: "",
                environment: arguments["environment"] as! String,
                action: arguments["action"] as! String,
                clientId: arguments["clientId"] as! String,
                transactionId: tranxId,
                merchantId: arguments["merchantId"] as! String,
                order: arguments["order"] as! [String: Any],
                billing: arguments["billing"] as! [String: Any],
                shipping: arguments["shipping"] as! [String: Any],
                appearance: arguments["appearance"] as! [String: Any]
            )
            
        } else {
            result(FlutterMethodNotImplemented)
        }
    }

    private let sonicManager: MCSonicController = MCSonicController()

    private func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
       
        UIFont.registerFontsForIzipayFramework() //Agregar fuentes al SDK
            
        return true
    }
    
    // Obtener la fecha y hora actual
    func getCurrentTimestamp() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
        return formatter.string(from: Date())
    }
    
    func calculateMillis(start: String, end: String) -> Int {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
        guard let startDate = formatter.date(from: start),
              let endDate = formatter.date(from: end) else {
            return 0
        }
        return Int(endDate.timeIntervalSince(startDate) * 1000)
    }

    public func getPaymentResult(_ paymentResult: IzipayPayButtonSDK.PaymentResult) {
    print("Código: \(paymentResult.code ?? "N/A")")
    print("Mensaje: \(paymentResult.message ?? "N/A")")

    let header: [String: Any] = [
        "transactionStartDatetime": transactionStartDatetime!,
        "transactionEndDatetime": getCurrentTimestamp(),
        "millis": calculateMillis(start: transactionStartDatetime!, end: getCurrentTimestamp())
    ]

    let card: [String: Any] = [
        "brand": paymentResult.response?.card?.brand ?? "N/A",
        "pan": paymentResult.response?.card?.pan ?? "N/A"
    ]

    let token: [String: Any] = [
        "cardToken": paymentResult.response?.token?.cardToken ?? "N/A"
    ]

    let response: [String: Any] = [
        "merchantCode": paymentResult.response?.merchant?.merchantCode ?? "N/A",
        "facilitatorCode": "",
        "merchantBuyerId": paymentResult.response?.token?.merchantBuyerId ?? "N/A",
        "card": card,
        "token": token,
        "transactionId": paymentResult.transactionId ?? "N/A",
        "orderNumber": paymentResult.response?.order?.first?.orderNumber ?? "N/A"
    ]

    let resultMap: [String: Any] = [
        "messageFriendly": paymentResult.message ?? "N/A"
    ]

    let dataMap: [String: Any] = [
        "code": paymentResult.code ?? "N/A",
        "message": paymentResult.message ?? "N/A",
        "header": header,
        "response": response,
        "result": resultMap
    ]

    var result = ""
    do {
        let jsonData = try JSONSerialization.data(withJSONObject: dataMap, options: .prettyPrinted)
        if let jsonString = String(data: jsonData, encoding: .utf8) {
            print(jsonString)
            result = jsonString
        }
    } catch {
        print("Error al convertir el diccionario a JSON: \(error.localizedDescription)")
    }

    if let paymentResultCode = paymentResult.code, paymentResultCode == "00" {
        if paymentResult.response != nil {
            print("Transacción exitosa. \(paymentResult.response?.description ?? "N/A")")
            sendResult(data: result, message: paymentResult.message)
        } else {
            sendResult(data: "", message: "Error: response is nil")
        }
    } else {
        print("Operación no completada o rechazada. Consultar/anular orden.")
        sendResult(data: result, message: paymentResult.message)
    }
}
    
    public func executeProfiling(_ params: IzipayPayButtonSDK.ScoringParams) {
        // Implementar según sea necesario
    }
    
    public func executeSensoryBrandVisa(view: UIView) {
        if let visaview = view as? SensoryBranding {
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
                visaview.animate()
            }
        }
    }

    private func sendResult(data: String = "", message: String? = nil) {
        channelResult([data])
    }
    
    public func executeSensoryBrandMastercard(view: UIView, params: IzipayPayButtonSDK.MastercardSonicParams) {
        let merchant = MCSonicMerchant(merchantName: params.name, countryCode: params.country,
                                       city: nil, merchantCategoryCodes: [params.mcc], merchantId: nil)
        if let mastercardview = view as? MCSonicView, let safeMerchant = merchant {
            self.sonicManager.prepare(with: .soundAndAnimation, sonicCue: params.sonicCue,
                                      sonicEnvironment: .production, merchant: safeMerchant) { status in
                if status == .success {
                    self.sonicManager.play(with: mastercardview) { MCSonicStatus in
                        print(MCSonicStatus)
                    }
                }
            }
        }
    }
    
    private func currentTimeMillis() -> Int64 {
        return Int64(Date().timeIntervalSince1970 * 1000 * 1000)
    }
    func showIzipayForm(
        listPayMethod: String,
        environment: String,
        action: String,
        clientId: String,
        transactionId: String,
        merchantId: String,
        order: [String: Any],
        billing: [String: Any],
        shipping: [String: Any],
        appearance: [String: Any]
    ) {
        let random = Int.random(in: 356547157..<366547157) //solo es un ejemplo, el transactionId debe ser unico y el comercio elige la forma de generarlo
        
        let tranxId = "\(random)"
        self.transactionId = tranxId
        
        var config = ConfigPaymentIzipay()
        config.enviroment = environment //TEST o PROD o SBOX
        config.merchantCode = merchantId
        config.facilitatorCode = ""
        config.publicKey = clientId
        
        config.transactionId = tranxId
        config.action = action
        
        config.order = OrderPaymentIzipay()
        config.order?.orderNumber = "10\(random)"
        config.order?.amount = order["amount"] as? String ?? "10.00"
        config.order?.currency = order["currency"] as? String ?? "PEN"
        //config.order?.channel = "mobile" ya no requerido
        config.order?.processType = "autorize" //"autorize" o "preautorize"
        config.order?.payMethod = [.all] //.all, .card
        config.order?.merchantBuyerId = "MB10\(random)" //es un ejemplo, debe ser unico por usuario del comercio
        config.order?.dateTimeTransaction = "\(currentTimeMillis())"
        config.billing = BillingPaymentIzipay()
        config.billing?.firstName = billing["firstName"] as? String ?? "John"
        config.billing?.lastName = billing["lastName"] as? String ?? "Doe"
        config.billing?.email = billing["email"] as? String ?? "ejemplo@gmail.com"
        config.billing?.phoneNumber = billing["phonphoneeNumber"] as? String ?? "989339999"
        config.billing?.street = billing["address"] as? String ?? "calle el demo sdk example"
        config.billing?.city = billing["city"] as? String ?? "lima"
        config.billing?.state = billing["region"] as? String ?? "lima"
        config.billing?.country = billing["country"] as? String ?? "PE"
        config.billing?.postalCode = billing["postalCode"] as? String ?? "15047"
        config.billing?.documentType = billing["idType"] as? String ?? "DNI"
        config.billing?.document = billing["idNumber"] as? String ?? "22334455"
        
        config.shipping = ShippingPaymentIzipay()
        config.shipping?.firstName = shipping["firstName"] as? String ?? "John"
        config.shipping?.lastName = shipping["lastName"] as? String ?? "Doe"
        config.shipping?.email = shipping["email"] as? String ?? "ejemplo@gmail.com"
        config.shipping?.phoneNumber = shipping["phone"] as? String ?? "989339999"
        config.shipping?.street = shipping["address"] as? String ?? "calle el demo sdk example"
        config.shipping?.city = shipping["city"] as? String ?? "lima"
        config.shipping?.state = shipping["region"] as? String ?? "lima"
        config.shipping?.country = shipping["country"] as? String ?? "PE"
        config.shipping?.postalCode = shipping["postalCode"] as? String ?? "15047"
        config.shipping?.documentType = shipping["idType"] as? String ?? "DNI"
        config.shipping?.document = shipping["idNumber"] as? String ?? "22334455"
        
        config.token = TokenPaymentIzipay()
        config.token?.cardToken = nil //requerido solo en action pay_token
        config.appearance = AppearencePaymentIzipay()
        config.appearance?.theme = appearance["themeColor"] as? String ?? "green"
        config.appearance?.logo = appearance["logoUrl"] as? String ?? "https://logowik.com/content/uploads/images/shopping-cart5929.jpg" //es solo un ejemplo
        config.appearance?.formControls = AppearenceControlsPaymentIzipay()
        config.appearance?.formControls?.isAmountLabelVisible = true
        config.appearance?.formControls?.isLangControlVisible = true
        config.appearance?.language = appearance["language"] as? String ?? "ESP"
        config.appearance?.visualSettings = AppearenceVisualSettingsPaymentIzipay()
        config.appearance?.visualSettings?.presentationMode = .fullscreen
        config.appearance?.visualSettings?.showResultScreen = true
        //Objetos para sensory branding
        let sensoryBrandingVisa = SensoryBranding()
        sensoryBrandingVisa.backdropColor = .white
        sensoryBrandingVisa.isSoundEnabled = true
        sensoryBrandingVisa.isHapticFeedbackEnabled = true
        sensoryBrandingVisa.checkmarkMode = .checkmark

        let sensoryBrandMastercard = MCSonicView()
        sensoryBrandMastercard.background = MCSonicBackground.white

        config.appearance?.sensoryBrand = AppearenceSensoryBrandIzipay()
        config.appearance?.sensoryBrand?.visaSBView = sensoryBrandingVisa
        config.appearance?.sensoryBrand?.mastercardSBView = sensoryBrandMastercard
        
        config.urlIPN = nil //enviar si se quiere notificaciones server to server
      
        
        let s = UIStoryboard(name: "IziPayment", bundle: Bundle(for: PaymentFormViewController.self))
        let vc = s.instantiateViewController(withIdentifier: "PaymentFormView") as! PaymentFormViewController
        vc.delegate = self
        vc.configPayment = config
        vc.modalPresentationStyle = .fullScreen

        // Obtener el rootViewController de la escena activa
         if let scene = UIApplication.shared.connectedScenes.first(where: { $0.activationState == .foregroundActive }),
            let windowScene = scene as? UIWindowScene,
            let rootViewController = windowScene.windows.first(where: { $0.isKeyWindow })?.rootViewController {

             // Presentar el PaymentFormViewController
            rootViewController.present(vc, animated: true, completion: nil)
        }
        
        
    }
}

File: android/src/main/kotlin/com/izipay/payment/izipay_payment/IzipayPaymentPlugin.kt
package com.izipay.payment.izipay_payment

import android.util.Log
import android.app.Activity
import android.content.Intent
import androidx.core.app.ActivityCompat
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.PluginRegistry
import com.izipay.izipay_pw_sdk.ui.fullscreend.ContainerActivity
import com.izipay.izipay_pw_sdk.data.model.*
import com.izipay.izipay_pw_sdk.domain.model.PaymentResponse

import org.json.JSONObject
import com.google.gson.Gson
import java.text.SimpleDateFormat
import java.util.*


/** IzipayPaymentPlugin */
class IzipayPaymentPlugin: FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware,
PluginRegistry.ActivityResultListener{
    private var channel: MethodChannel? = null
    private var channelResult: MethodChannel.Result? = null
    private var activity: FlutterActivity? = null
    private var flutterBinding: FlutterPlugin.FlutterPluginBinding? = null
    private var activityBinding: ActivityPluginBinding? = null
    private var transactionStartDatetime: String = ""

    private lateinit var listPayMethod: ArrayList<PayOption>

    companion object {
        const val CHANNEL_NAME = "izipay_payment_flutter"
        const val EXPECTED_REQUEST_CODE = 1001
        

        // @JvmStatic
        // fun registerWith(registrar: PluginRegistry.Registrar) {
        //     val channel = MethodChannel(registrar.messenger(), CHANNEL_NAME)
        //     channel.setMethodCallHandler(IzipayPaymentPlugin())
        // }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        channel?.setMethodCallHandler(this)
        flutterBinding = flutterPluginBinding
    }

    override fun onDetachedFromEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        flutterBinding = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        channel = MethodChannel(flutterBinding!!.binaryMessenger, CHANNEL_NAME)
        channel?.setMethodCallHandler(this)

        activityBinding = binding
        activityBinding?.addActivityResultListener(this)
        activity = binding.activity as FlutterActivity
    }

    override fun onDetachedFromActivity() {
        channel?.setMethodCallHandler(null)
        channel = null
        channelResult = null

        activityBinding?.removeActivityResultListener(this)
        activityBinding = null
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method == "startPayment") {
            transactionStartDatetime = getCurrentTimestamp()
            channelResult = result

            val args = call.arguments as? Map<String, Any> ?: run {
                result.error("INVALID_ARGUMENTS", "Invalid arguments received from Dart", null)
                return
            }

            val environment = args["environment"] as? String ?: "SBOX"
            val action = args["action"] as? String ?: "register"
            val clientId = args["clientId"] as? String ?: ""//"<CODIGO DE COMERCIO>"
            val merchantId = args["merchantId"] as? String ?: ""//<public key>
            val orderTimestamp = System.currentTimeMillis().toString()

            listPayMethod = arrayListOf(PayOption.CARD)
            Log.d("IzipayPaymentPlugin", "Processing payment: $args")

            val request = buildConfigRequest(
                listPayMethod,
                environment,
                action,
                clientId,
                "DMKTL$orderTimestamp",
                merchantId,
                orderTimestamp,
                args["order"] as? Map<String, Any> ?: mapOf(),
                args["billing"] as? Map<String, Any> ?: mapOf(),
                args["shipping"] as? Map<String, Any> ?: mapOf(),
                args["appearance"] as? Map<String, Any> ?: mapOf()
            )
            Log.d("IzipayPaymentPlugin", "Processing payment buildConfigRequest: $request")

            val intent = Intent(activity, ContainerActivity::class.java).apply {
                putExtra(ContainerActivity.REQUEST, request)
            }
            activity?.startActivityForResult(intent, EXPECTED_REQUEST_CODE)
        } else {
            result.notImplemented()
        }
    }

    fun getCurrentTimestamp(): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.getDefault())
        return formatter.format(Date())
    }
    
    fun calculateMillis(start: String, end: String): Int {
        val formatter = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.getDefault())
        return try {
            val startDate = formatter.parse(start)
            val endDate = formatter.parse(end)
            if (startDate != null && endDate != null) {
                (endDate.time - startDate.time).toInt()
            } else {
                0
            }
        } catch (e: Exception) {
            0
        }
    }

    private fun buildConfigRequest(
        listPayMethod: ArrayList<PayOption>,
        environment: String,
        action: String,
        clientId: String,
        transactionId: String,
        merchantId: String,
        timestamp: String,
        order: Map<String, Any>,
        billing: Map<String, Any>,
        shipping: Map<String, Any>,
        appearance: Map<String, Any>
    ): ConfigRequest {
        val orderTimestamp = System.currentTimeMillis().toString()
        return ConfigRequest(
            environment,
            action,
            clientId,
            transactionId,
            merchantId,
            "",
            OrderPaymentIzipay(
                timestamp,
                order["currency"] as? String ?: "PEN",
                order["amount"] as? String ?: "11.00",
                listPayMethod,
                "mobile",
                "autorize",
                "MB10$orderTimestamp",
                timestamp
            ),
            TokenPaymentIzipay(""),
            BillingPaymentIzipay(
                billing["firstName"] as? String ?: "Juan",
                billing["lastName"] as? String ?: "Quispe",
                billing["email"] as? String ?: "ejemplo01@gmail.com",
                billing["phone"] as? String ?: "930292619",
                billing["address"] as? String ?: "Av. flores",
                billing["city"] as? String ?: "Lima",
                billing["region"] as? String ?: "Lima",
                billing["country"] as? String ?: "PE",
                billing["postalCode"] as? String ?: "33065",
                billing["idType"] as? String ?: "DNI",
                billing["idNumber"] as? String ?: "55555555"
            ),
            ShippingPaymentIzipay(
                shipping["firstName"] as? String ?: "Juan",
                shipping["lastName"] as? String ?: "Quispe",
                shipping["email"] as? String ?: "ejemplo01@gmail.com",
                shipping["phone"] as? String ?: "930292619",
                shipping["address"] as? String ?: "Av. Alamo",
                shipping["city"] as? String ?: "Lima",
                shipping["region"] as? String ?: "Lima",
                shipping["country"] as? String ?: "PE",
                shipping["postalCode"] as? String ?: "33065",
                shipping["idType"] as? String ?: "DNI",
                shipping["idNumber"] as? String ?: "55555555"
            ),
            AppearencePaymentIzipay(
                appearance["language"] as? String ?: "ESP",
                AppearenceControlsPaymentIzipay(true, true),
                AppearenceVisualSettingsPaymentIzipay(true),
                "green",
                CustomThemePaymentIzipay(
                    appearance["primaryColor"] as? String ?: "#333399",
                    appearance["secondaryColor"] as? String ?: "#333399",
                    appearance["tertiaryColor"] as? String ?: "#333399"
                ),
                appearance["logoUrl"] as? String ?: "https://logowik.com/content/uploads/images/shopping-cart5929.jpg"
            ),
            ""
        )
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode == EXPECTED_REQUEST_CODE) {
            val dataPayLoad = data?.extras?.getString(ContainerActivity.RESPONSEPAYLOAD)
            
            val jsonData = JSONObject(dataPayLoad)
            val response = jsonData.optJSONObject("response")
            val merchant = response?.optJSONObject("merchant")
            val card = response?.optJSONObject("card")
            val token = response?.optJSONObject("token")
            val order = response?.optJSONArray("order")?.optJSONObject(0)
            val dataMap = mapOf(
                "code" to jsonData.optString("code"),
                "message" to jsonData.optString("message"),
                "header" to mapOf(
                    "transactionStartDatetime" to transactionStartDatetime,
                    "transactionEndDatetime" to getCurrentTimestamp(),
                    "millis" to calculateMillis(transactionStartDatetime, getCurrentTimestamp())
                ),
                "response" to if (response != null) {
                    mapOf(
                        "merchantCode" to (merchant?.optString("merchantCode", "") ?: ""),
                        "facilitatorCode" to "",
                        "merchantBuyerId" to (token?.optString("merchantBuyerId", "") ?: ""),
                        "card" to mapOf(
                            "brand" to (card?.optString("brand", "") ?: ""),
                            "pan" to (card?.optString("pan", "") ?: "")
                        ),
                        "token" to mapOf(
                            "cardToken" to (token?.optString("cardToken", "") ?: "")
                        ),
                        "transactionId" to (jsonData?.optString("transactionId", "") ?: ""),
                        "orderNumber" to (order?.optString("orderNumber", "") ?: ""),
                    )
                } else {
                    emptyMap() // Si no hay respuesta, dejar vacío
                },
                "payload" to jsonData.optString("payloadHttp", ""),
                "result" to mapOf(
                    "messageFriendly" to jsonData.optString("messageUser", "")
                )
            )

            val gson = Gson()
            val jsonGson = gson.toJson(dataMap)

            if (resultCode == Activity.RESULT_OK) {
                channelResult?.success(jsonGson)
            } else {
                channelResult?.error("PAYMENT_FAILED", "$jsonGson", null)
            }
            return true
        }
        return false
    }
}
